src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -4035,10 +4037,18 @@ private void ParseParameterModifiers(SyntaxListBuilder modifiers)
                 {
                     case SyntaxKind.ThisKeyword:
                         modifier = CheckFeatureAvailability(modifier, MessageID.IDS_FeatureExtensionMethod);
+                        if (seenRefModifier)
+                        {
+                            modifier = CheckFeatureAvailability(modifier, MessageID.IDS_FeatureRefExtensionMethods);
```

Comment:
I think CheckFeatureAvailability error is pretty clean and friendly enough. And there is a convenience that UpgradeProject automatically works. 

Most of these errors are point-in-time issues anyways. New version becomes the default fairly quickly. - you rarely see "'async' requires version X" nowdays.